drivers/include/eeprom.h

```diff
@@ -0,0 +1,194 @@
+
+/**
+ * @brief   Write a byte at the given position
+ *
+ * @param[in]       dev     Device handle of the EEPROM device to read from
+ * @param[in]       pos     Position to write to
+ * @param[in]       data    Data to write
+ *
+ * @retval  0               Success
+ * @retval  -ENODEV         No device for given handle in @p dev
+ * @retval  -EFAULT         Bad address given in @p pos
+ * @retval  -EINVAL         Parameter @p dest is `NULL`
+ */
+static inline int eeprom_write_byte(eeprom_t dev, eeprom_off_t pos,
+                                    uint8_t data);
+
+/**
+ * @brief   Clear @p len bytes from the given position @p pos
+ *
+ * @param[in]   dev     Device handle of the EEPROM device to clear data from
+ * @param[in]   pos     Position of the first byte to clear
+ * @param[in]   len     The number of bytes to clear
+ *
+ * @retval  0           Success
+ * @retval  -ENODEV     No device for given handle in @p dev
+ * @retval  -EFAULT     Bad address given in @p pos
+ * @retval  -EINVAL     Parameter @p dest is `NULL`
+ * @retval  -EIO        I/O error while writing to the device
+ *
+ * Clearing a byte in EEPROM simply consists in setting it to 0
+ *
+ * @details If this functions fails with `-EIO`, the contents in the range
+ *          `[pos; pos + len[` on the device are undefined. For every other
+ *          error the EEPROM is not changed.
+ */
+static inline int eeprom_clear(eeprom_t dev, eeprom_off_t pos, size_t len);
+
+/**
+ * @brief   Registers an EEPROM device to the common API
+ *
+ * @param[in]   driver  Driver of the EEPROM device
+ * @param[in]   handle  Handle to pass to the driver
+ * @param[in]   size    Size of the EEPROM device to add in bytes
+ *
+ * @pre     This function must be called once for each device
+ * @pre     This function must not be called more than @ref EEPROM_NUMOF
+ *          times
+ */
+void eeprom_register(const eeprom_driver_t *driver, void *handle,
+                     eeprom_off_t size);
+
+/**
+ * @brief   Set @p len bytes from the given position @p pos with value @p val
+ *
+ * @param[in]   dev     Device handle of the EEPROM device to write to
+ * @param[in]   pos     Start position in EEPROM
+ * @param[in]   val     The value to set
+ * @param[in]   len     The number of bytes to set
+ *
+ * @retval  0           Success
+ * @retval  -ENODEV     No device for given handle in @p dev
+ * @retval  -EFAULT     Bad address given in @p pos
+ * @retval  -EINVAL     Parameter @p dest is `NULL`
+ * @retval  -EIO        I/O error while writing to the device
+ *
+ * @details If this functions fails with `-EIO`, the contents in the range
+ *          `[pos; pos + len[` on the device are undefined. For every other
+ *          error the EEPROM is not changed.
+ */
+int eeprom_set(eeprom_t dev, eeprom_off_t pos, uint8_t val, size_t len);
+
+#ifdef __cplusplus
+}
+#endif
+
+/* Include implementation of the static inline functions */
+#include "eeprom_implementation.h"
```

Comment:
Is there a real benefit with the static inline functions ? Is not I would just implement declare them here and implement them in `eeprom.c`. Or maybe keep them static inline but put them in `eeprom.h`.

It's just that I don't like having to include a header at the end of another one so it's better if we could avoid this.

Reply:
> Is there a real benefit with the static inline functions?

It depends. In any case, it will be faster. In some cases, the compiler will be able to optimize out (some of) the range checks done there.

Regarding ROM size, this could be both a minor improvement or a minor regression. It will depend on the number of places where the common API is called and on how much of the code can be optimized out. In general, I expect small applications to be slightly small with this, and other being slightly bigger.

I have no strong feelings about this; so I'd just as happy move it back to `eeprom.c`.

@benpicco suggested to reduce the overhead here, maybe he wants to comment before I move that back?

> It's just that I don't like having to include a header at the end of another one so it's better if we could avoid this.

This could actually also be included at the top of the file, if no compiler flags are set to warn about late declarations.  But I think the C standard is fine with multiple or late declarations (as long as declarations and the implementations match in name and signature). The advantage of still having the declaration here is that the generated documentation will be properly grouped. And the advantage of moving the implementation out is the separation of API and implementation.

This is by the way also not uncommon, e.g. [`xtimer.h`](https://github.com/RIOT-OS/RIOT/blob/master/sys/include/xtimer.h) includes the static inline implementations at the end of the header.